src/data/mutable_data.rs

```diff
@@ -169,59 +180,117 @@ impl MutableData {
         &self.data
     }
 
-    /// Inserts a new entry (key + value pair)
-    pub fn ins_entry(&mut self,
-                     key: Vec<u8>,
-                     value: Value,
-                     requester: PublicKey)
-                     -> Result<(), RoutingError> {
-        if !self.is_action_allowed(requester, Action::Insert) {
-            return Err(RoutingError::AccessDenied);
-        }
-        if self.data.contains_key(&key) {
-            return Err(RoutingError::EntryAlreadyExist);
+    fn rollback(&mut self,
+                inserted: BTreeSet<Vec<u8>>,
+                updated: BTreeMap<Vec<u8>, Value>,
+                deleted: BTreeMap<Vec<u8>, Value>) {
+        for (key, val) in deleted {
+            let _ = self.data.insert(key, val);
         }
-        if self.data.len() > MAX_MUTABLE_DATA_ENTRIES as usize {
-            return Err(RoutingError::TooManyEntries);
+        for (key, val) in updated {
+            let _ = self.data.insert(key, val);
         }
-        let _ = self.data.insert(key.clone(), value);
-        if !self.validate_mut_size() {
+        for key in inserted {
             let _ = self.data.remove(&key);
-            return Err(RoutingError::ExceededSizeLimit);
         }
-        Ok(())
     }
 
-    /// Updates an existing entry (key + value pair)
-    pub fn update_entry(&mut self,
-                        key: Vec<u8>,
-                        value: Value,
-                        requester: PublicKey)
-                        -> Result<(), RoutingError> {
-        if !self.is_action_allowed(requester, Action::Update) {
+    /// Mutates entries (key + value pairs) in bulk
+    pub fn mutate_entries(&mut self,
+                          actions: BTreeMap<Vec<u8>, EntryAction>,
+                          requester: PublicKey)
+                          -> Result<(), RoutingError> {
+        // Deconstruct actions into inserts, updates, and deletes
+        let (insert, update, delete) = actions.into_iter()
+            .fold((BTreeMap::new(), BTreeMap::new(), BTreeMap::new()),
+                  |(mut insert, mut update, mut delete), (key, item)| {
+                match item {
+                    EntryAction::Ins(value) => {
+                        let _ = insert.insert(key, value);
+                    }
+                    EntryAction::Update(value) => {
+                        let _ = update.insert(key, value);
+                    }
+                    EntryAction::Del(version) => {
+                        let _ = delete.insert(key, version);
+                    }
+                };
+                (insert, update, delete)
+            });
+
+        if (insert.len() > 0 && !self.is_action_allowed(requester, Action::Insert)) ||
```

Comment:
https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.is_empty is more idiomatic i think instead of `len() > 0`